### ISSUE TITLE
Error while performing analysis because of type recursion

### ISSUE THREAD

**Issue by sintell** [microsoft/pyright#4535]
**Describe the bug**
Type analysator fails to compute type for the following function:
```py
def function_with_huge_result_type(k):
    keys =  ["1", "2", "b"]
    value = []

    while keys:
        if not k:
            continue

        if not k:
            value = {k: value}
        else:
            value = [None] * int(k) + [value]

    return value
```

The traceback I got is:
```
"Error performing analysis: RangeError: Invalid string length
    at Array.join (<anonymous>)
    at Object.printType (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typePrinter.ts:516:63)
    at printType (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:25261:28)
    at assignFromUnionType (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:22912:33)
    at assignType (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:22201:20)
    at getTypeOfExpression (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:1206:22)
    at forEach (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:13203:35)
    at Array.forEach (<anonymous>)
    at getTypeOfListOrSetWithContext (/Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:13197:22)
    at /Users/antky/.local/share/nvim/mason/packages/pyright/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:13135:28"
```

Even if I narrow down types recursion, depending on the type of change, it is possible that pyright still will be computing resulting type for a while, but will complete computation without errors.

**To Reproduce**
1. Create a file containing following code:
```py
def function_with_huge_result_type(k):
    keys =  ["1", "2", "b"]
    value = []

    while keys:
        if not k:
            continue

        if not k:
            value = {k: value}
        else:
            value = [None] * int(k) + [value]

    return value
```
2. Run pyright binary on it
3. Get the error

**Expected behavior**
I expect pyright to limit recursion depth somehow.
It'll also be great to have an option to skip type computation if it takes too long

**Screenshots or Code**
See above

**VS Code extension or command-line**
I use command-line version as a part of my vim LSP setup
The version is `pyright 1.1.291`

**Comment by sintell** [comment]
My issue is definitely related with this one on the cause: https://github.com/microsoft/pyright/issues/4519
But I can't tell if the fix you've implemented for the above will also help with my issue

**Comment by erictraut** [comment]
Thanks for the bug report. This isn't related to #4519. It is related to #4424, which I recently fixed for tuples. The same fix should be applied for lists, sets, and dicts. This will be addressed in the next release.

**Comment by sintell** [comment]
Great, thanks!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.292, which I just published. It will also be included in a future release of pylance.
